src/bin/pgcopydb/ld_transform.c

```diff
@@ -180,14 +180,18 @@ stream_transform_line(void *ctx, const char *line, bool *stop)
 		}
 
 		/* now write the transaction out */
-		if (!stream_write_message(privateContext->out, currentMsg))
+		if (!stream_write_message(privateContext,
+								  privateContext->out,
+								  currentMsg))
 		{
 			/* errors have already been logged */
 			return false;
 		}
 
 		/* now write the transaction out also to file on-disk */
-		if (!stream_write_message(privateContext->sqlFile, currentMsg))
+		if (!stream_write_message(privateContext,
+								  privateContext->sqlFile,
+								  currentMsg))
```

Comment:
I didn't expect this to change in the context of your PR. I believe that's the place where we should consider also writing the xid.json file, when the message is the STREAM_ACTION_COMMIT of a continued transaction. Otherwise there is this whole business of distinguishing between writing to a file or to stdout, and we don't want the lower-level API to know about that, all they know about is a FILE * output stream.

Reply:
Thank you, I think this way I can also eliminate similar problem in `ld_apply.c` and fix a problem that I left as a follow-up inside this PR only.